### PR TITLE
Grouped product frontend quantity validation added and code refactor

### LINKED ISSUE #39142: Magento 2.4.7 minAllowed missing product order qty
### Preconditions and environment

- Magento version 2.4.7-P2
- See source code product page

### Steps to reproduce

1. View product (on product page)
2. View Source 
3. Search (CTRL + F) maxAllowed

### Expected result

input type="number" name="qty" id="qty" maxlength="12" value="0.5" title="Aantal" class="input-text qty" data-validate="{&quot;required-number&quot;:true,&quot;validate-item-quantity&quot;:{&quot;**minAllowed**&quot;:0.5,&quot;maxAllowed&quot;:10000}}" /

### Actual result

input type="number" name="qty" id="qty" min="0" value="0.6" title="Aantal" class="input-text qty" data-validate="{&quot;required-number&quot;:true,&quot;validate-item-quantity&quot;:{&quot;maxAllowed&quot;:999}}" /

### Additional information

min="0" is present, but not correct (min order qty is 0.6 for this product)

### Release note

_No response_

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [X] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### LINKED ISSUE #39479: Grouped Product Frontend Quantity Validations are Missing
### Preconditions and environment

- Magento version 2.4.7-p3

### Steps to reproduce

1. Create a Grouped Product and associate it with child products.
2. Set the following quantity properties for one of the child products:

    - Minimum Qty Allowed in Shopping Cart: 2
    - Maximum Qty Allowed in Shopping Cart: 10
    - Qty Increments: 2

3.Open the Product Detail Page (PDP) of the Grouped Product and test invalid quantities for the selected child product (e.g., 1, 3, or 11) to confirm that invalid inputs are restricted.

### Expected result

The validation error should be displayed upon clicking the "Add to Cart" button, without reloading the page, similar to the behavior on the Simple Product PDP.

image

### Actual result

Instead of displaying the error immediately, the page reloads and the error is shown afterward from the server side.

image

### Additional information

_No response_

### Release note

_No response_

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by Mohamed-Asar**
### Description (*)
Grouped Product Frontend Quantity validation added

### Fixed Issues (if relevant)
1. #39479 
2. #39142

### Manual testing scenarios (*)

1.Create a Grouped Product and associate it with child products.
2. Set the following quantity properties for one of the child products:
    - Minimum Qty Allowed in Shopping Cart: 2
    - Maximum Qty Allowed in Shopping Cart: 10
    - Qty Increments: 2
 3.Open the Product Detail Page (PDP) of the Grouped Product and test invalid quantities for the selected child product (e.g., 1, 3, or 11) to confirm that invalid inputs are restricted.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by Mohamed-Asar** [comment]
@magento run all tests.

**Comment by Mohamed-Asar** [comment]
@magento run all tests

**Comment by Mohamed-Asar** [comment]
@magento run all tests

**Comment by Mohamed-Asar** [comment]
@magento run all tests

**Comment by Mohamed-Asar** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

File: app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
+<?php
+/**
+ * Copyright 2024 Adobe
+ * All Rights Reserved.
+ */
+declare(strict_types=1);
+
+namespace Magento\CatalogInventory\Model\Product;
+
+use Magento\CatalogInventory\Api\StockRegistryInterface;
```
Comment (engcom-Hotel) [inline app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php:10]: interface StockRegistryInterface has been deprecated. Please use alternate.

File: app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    /**
+     * @var StockRegistryInterface
+     */
+    private $stockRegistry;
+
+    /**
+     * @param StockRegistryInterface $stockRegistry
+     */
+    public function __construct(
+        StockRegistryInterface $stockRegistry
```
Comment (engcom-Hotel) [inline app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php]: Use `readonly` instead. Please refer to [this](https://github.com/magento/magento2/blob/3428ebc3802a26fae9be542163579cefaa7967fa/app/code/Magento/Tax/Model/Plugin/AddTaxesExtensionAttribute.php#L46).

File: app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    /**
+     * To get quantity validators
+     *
+     * @param int $productId
+     * @param int $websiteId
+     *
+     * @return array
+     */
+    public function getData($productId, $websiteId): array
```
Comment (engcom-Hotel) [inline app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php]: The parameter's type declaration is missing. Please add.

File: app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        $stockItem = $this->stockRegistry->getStockItem($productId, $websiteId);
+
+        $params = [];
+        $validators = [];
+        $params['minAllowed'] = (float)$stockItem->getMinSaleQty();
```
Comment (engcom-Hotel) [inline app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php]: `getMinSaleQty` is returning the `float` value. No need to typecast here.

File: app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        if ($stockItem->getMaxSaleQty()) {
+            $params['maxAllowed'] = (float)$stockItem->getMaxSaleQty();
```
Comment (engcom-Hotel) [inline app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php]: Same here

File: app/code/Magento/GroupedProduct/ViewModel/ValidateQuantity.php [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+     * @var QuantityValidator
+     */
+    private $productQuantityValidator;
+
+    /**
+     * @param Json $serializer
+     * @param QuantityValidator $productQuantityValidator
+     */
+    public function __construct(
+        Json $serializer,
```
Comment (engcom-Hotel) [inline app/code/Magento/GroupedProduct/ViewModel/ValidateQuantity.php]: Use `readonly` constructor parameter.

File: app/code/Magento/GroupedProduct/ViewModel/ValidateQuantity.php [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+    /**
+     * To get the quantity validators
+     *
+     * @param int $productId
+     * @param int $websiteId
+     *
+     * @return string
+     */
+    public function getQuantityValidators($productId, $websiteId): string
```
Comment (engcom-Hotel) [inline app/code/Magento/GroupedProduct/ViewModel/ValidateQuantity.php]: Please add parameter's type declaration.

File: app/code/Magento/CatalogInventory/Block/Plugin/ProductView.php [hunk 8]
```diff
@@ -1,26 +1,26 @@
...
-    private $stockRegistry;
+    private $productQuantityValidator;
 
     /**
-     * @param StockRegistryInterface $stockRegistry
+     * @param QuantityValidator $productQuantityValidator
      */
     public function __construct(
-        StockRegistryInterface $stockRegistry
+        QuantityValidator $productQuantityValidator
```
Comment (engcom-Hotel) [inline app/code/Magento/CatalogInventory/Block/Plugin/ProductView.php:22]: Use ObjectManager to initialize as this change is BiC.

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Functional Tests CE, Static Tests, Unit Tests

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @Mohamed-Asar,

Thanks for the contribution!

Please find below the review comments.

Thanks

File: app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php (hunk 1 above)
Comment (Mohamed-Asar) [reply app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php:10]: @engcom-Hotel The alternative for that interface is the MSI module. However, some users might have the MSI module disabled, can i use that ?

File: app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php (hunk 2 above)
Comment (Mohamed-Asar) [reply app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php]: done.

File: app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php (hunk 3 above)
Comment (Mohamed-Asar) [reply app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php]: done

File: app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php (hunk 4 above)
Comment (Mohamed-Asar) [reply app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php]: removed the typecast

File: app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php (hunk 5 above)
Comment (Mohamed-Asar) [reply app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php]: removed the typecast

File: app/code/Magento/GroupedProduct/ViewModel/ValidateQuantity.php (hunk 6 above)
Comment (Mohamed-Asar) [reply app/code/Magento/GroupedProduct/ViewModel/ValidateQuantity.php]: done

File: app/code/Magento/GroupedProduct/ViewModel/ValidateQuantity.php (hunk 7 above)
Comment (Mohamed-Asar) [reply app/code/Magento/GroupedProduct/ViewModel/ValidateQuantity.php]: done

File: app/code/Magento/CatalogInventory/Block/Plugin/ProductView.php (hunk 8 above)
Comment (Mohamed-Asar) [reply app/code/Magento/CatalogInventory/Block/Plugin/ProductView.php:22]: done

File: app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php (hunk 1 above)
Comment (engcom-Hotel) [reply app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php:10]: Yes @Mohamed-Asar you can use that. BTW its not recommended to disable the MSI module.

**Comment by engcom-Hotel** [comment]
Moving back to `Changes requested` as 1 [review comment](https://github.com/magento/magento2/pull/39480#discussion_r1926716718) still needs to be worked on.

File: app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php (hunk 1 above)
Comment (Mohamed-Asar) [reply app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php:10]: @engcom-Hotel MSI module is not included in this repo. how do i proceed ?

Comment (engcom-Hotel) [reply app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php:10]: Please refer to this repo: https://github.com/magento/inventory

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by Mohamed-Asar** [comment]
@engcom-Hotel  Not able to install MSI module, facing the below issue 

image

This issue occurs due to recent changes, which was already discussed [here](https://magentocommeng.slack.com/archives/C7KB93M32/p1736933306053669), i will proceed once the issue resolved

**Comment by engcom-Hotel** [comment]
Sure @Mohamed-Asar, until we are moving the PR `On Hold`.

Thanks

**Comment by Mohamed-Asar** [comment]
> @engcom-Hotel Not able to install MSI module, facing the below issue
> 
> image
> 
> This issue occurs due to recent changes, which was already discussed [here](https://magentocommeng.slack.com/archives/C7KB93M32/p1736933306053669), i will proceed once the issue resolved

@engcom-Hotel This issue was fixed by internal team, i will start working on the changes and let you know

File: app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php (hunk 1 above)
Comment (Mohamed-Asar) [reply app/code/Magento/CatalogInventory/Model/Product/QuantityValidator.php:10]: @engcom-Hotel I used MSI and remvoed the deprecated methods, please review

**Comment by Mohamed-Asar** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @Mohamed-Asar,

Thanks for updating the PR as per the review comment!

It seems some tests are still failing due to the changes. Please look into those.

Thanks

**Comment by Mohamed-Asar** [comment]
@magento run all tests

**Comment by Mohamed-Asar** [comment]
> Hello @Mohamed-Asar,
> 
> Thanks for updating the PR as per the review comment!
> 
> It seems some tests are still failing due to the changes. Please look into those.
> 
> Thanks

@engcom-Hotel Do i need to add the depency in composer.json for this modules( [Magento\InventoryConfigurationApi, Magento\InventorySalesApi]) ? and also these modules are not on this repo, so how can i proceed, do i need to mention in root composer.json and the module composer.json ?

**Comment by engcom-Hotel** [comment]
@magento run Static Tests

**Comment by engcom-Hotel** [comment]
Hello @Mohamed-Asar,

I guess you need to mention the dependencies in the `composer.json` of `magento/module-catalog-inventory` module here:
https://github.com/Mohamed-Asar/magento2/blob/559d0f8314f93b4aa59562d1a171329242dcffdb/app/code/Magento/CatalogInventory/composer.json#L7-L16

Thanks

**Comment by Mohamed-Asar** [comment]
@engcom-Hotel i added before, static test failed due to composer dependency issued, please see this [commit](https://github.com/magento/magento2/pull/39480/commits/baeb177a3bdf99a46a8744ee943f6146c291df9b)

Screenshot from 2025-04-07 15-56-58

**Comment by engcom-Hotel** [comment]
@Mohamed-Asar I suggest you to please go through with this documentation: https://developer.adobe.com/commerce/php/architecture/modules/dependencies/

I think we need to add them back.

Thanks

**Comment by Mohamed-Asar** [comment]
@magento run all tests

**Comment by Mohamed-Asar** [comment]
@engcom-Hotel I’ve added the module dependency in the CatalogInventory module's composer.json, but the static tests are still failing. It seems we need to add the dependency in the root composer.json file. Would it be okay if I go ahead and add it there as well? Please let me know your suggestion.

Also, I noticed that the following [MFTF test](https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Catalog/Test/Mftf/Test/AdminAddInStockProductToTheCartTest.xml) is consistently failing. I reviewed the test case and found that it sets min_qty as 4, but tries to add only 1 quantity to the cart—expecting the product to be added successfully.

However, in this PR, I’ve included min_qty validation, which is why the test is failing. Would it be okay if I update the test case accordingly?

Thanks!

**Comment by engcom-Hotel** [comment]
Hello @Mohamed-Asar,

Regarding the MFTF test failure, yes please update the test. As you resolved the bug related to validation which was not in place previously it might happen some existing tests fails. So generally we need to fix those.

And regarding the composer dependency static test failure we will get back to you soon on this.

Thanks

**Comment by Mohamed-Asar** [comment]
@magento run all tests

**Comment by Mohamed-Asar** [comment]
Hi @engcom-Hotel

I fixed the mftf test case for min qty validation

**Comment by Mohamed-Asar** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Integration Tests, Static Test

**Comment by Mohamed-Asar** [comment]
@magento run Static Tests, Functional Tests B2B

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B, Static Tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

File: app/code/Magento/CatalogInventory/Block/Plugin/ProductView.php (hunk 8 above)
Comment (engcom-Hotel) [reply app/code/Magento/CatalogInventory/Block/Plugin/ProductView.php:22]: not done yet, still cant see ObjectManager usages

File: app/code/Magento/CatalogInventory/Test/Unit/Model/Product/QuantityValidatorTest.php [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+use Magento\CatalogInventory\Api\StockRegistryInterface;
+use Magento\CatalogInventory\Model\Product\QuantityValidator;
+use Magento\CatalogInventory\Api\Data\StockItemInterface;
+use PHPUnit\Framework\MockObject\MockObject;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * @covers \Magento\CatalogInventory\Model\Product\QuantityValidator
+ */
+class QuantityValidatorTest extends TestCase
```
Comment (engcom-Hotel) [inline app/code/Magento/CatalogInventory/Test/Unit/Model/Product/QuantityValidatorTest.php:19]: Covering some edge case tests to ensure the code behaves correctly under unusual or extreme conditions. This could include scenarios like:

- `getMinSaleQty`, `getMaxSaleQty`, or `getQtyIncrements` returning null or invalid values.
- Testing with a `productId` or `websiteId` that does not exist in the stock registry.

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @Mohamed-Asar,
Thanks for the contribution!

Please have a look into the below review comment and fix the failed tests.

Thanks

File: app/code/Magento/CatalogInventory/Block/Plugin/ProductView.php (hunk 8 above)
Comment (Mohamed-Asar) [reply app/code/Magento/CatalogInventory/Block/Plugin/ProductView.php:22]: @engcom-Hotel i tried that, but the static test failed due to the object being initialized as null, so I had to revert the changes.

image

**Comment by engcom-Hotel** [comment]
Yes please @Mohamed-Asar, let us know of you are facing any difficulty in resolving the same.

**Comment by Mohamed-Asar** [comment]
@magento run all tests

**Comment by Mohamed-Asar** [comment]
> Hello @Mohamed-Asar, Thanks for the contribution!
> 
> Please have a look into the below review comment and fix the failed tests.
> 
> Thanks

@engcom-Hotel  If i added null i'm getting this error thats why i reverted it 

image

**Comment by Mohamed-Asar** [comment]
> Hello @Mohamed-Asar,
> 
> If you are facing issue in resolving the static test failures, then we can check the possibilities of creating JIRA for the approval.
> 
> Please let us know.
> 
> Thanks

@engcom-Hotel  Let me check and provide you an update by Monday

**Comment by Mohamed-Asar** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Review by engcom-Hotel — APPROVED** [review]

**Comment by engcom-Bravo** [comment]
Hi @Mohamed-Asar,

Thanks for your Contribution!!.

Could you please resolve the Conflicts to proceed further.

Thanks.

**Comment by Mohamed-Asar** [comment]
@engcom-Bravo I fixed the conflicts, please check

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Review by engcom-Hotel — APPROVED** [review]

**Comment by engcom-Dash** [comment]
@magento run all tests

**Comment by engcom-Dash** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Static Tests, Unit Tests

**Comment by engcom-Dash** [comment]
The consistent test failures for Functional B2B are known Issues and JIRA is raised for them.
Build 1:  https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39480/9ca147ac593cebfa8bb15cddfc34b291/Functional/allure-report-b2b/index.html#categories/8fb3a91ba5aaf9de24cc8a92edc82b5d

image

Build 2: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39480/24ce85251ad7e859543698323eef8b45/Functional/allure-report-b2b/index.html#categories/8fb3a91ba5aaf9de24cc8a92edc82b5d

image

Known Issues:
EnablePaypalExpressCheckoutAndSubmitAnOrderUsingPaypalExpressCheckoutTest : [ACQE-8007](https://jira.corp.adobe.com/browse/ACQE-8007)
StoreFrontSimpleProductWithSpecialAndTierDiscountPriceTest : [ACQE-7971](https://jira.corp.adobe.com/browse/ACQE-7971)
OrderWithMultipleConfigurableProductsAdditionalStockTest : [ACQE-8060](https://jira.corp.adobe.com/browse/ACQE-8060)

The consistent test failures for Functional CE is known Issues and JIRA is raised for the same.

Build 1: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39480/ea2a8bbe902a91cd403d34cc0f1413f7/Functional/allure-report-ce/index.html#categories/8fb3a91ba5aaf9de24cc8a92edc82b5d

image

Build 2: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39480/c6141d1ffdcb2b88bd9059b81502ebd3/Functional/allure-report-ce/index.html#categories/8fb3a91ba5aaf9de24cc8a92edc82b5d

image

Known Issues:
EnablePaypalExpressCheckoutAndSubmitAnOrderUsingPaypalExpressCheckoutTest : [ACQE-8007](https://jira.corp.adobe.com/browse/ACQE-8007)


The consistent test failure for Functional EE is known Issues and JIRA is raised for the same. Other failures are inconsistent and seems to be flaky. They neither part of PR nor failing because of the PR changes.

Build 1: https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39480/213c9121f0385ecc0ad9b0794b6719f8/Functional/allure-report-ee/index.html#categories/8fb3a91ba5aaf9de24cc8a92edc82b5d

image

Build 2:https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39480/60b1d7f04550acbc39bb8209e303e64c/Functional/allure-report-ee/index.html#categories/34dd2aa3ce3269b0c73667a1e007a61f/7ad1551983068770/

image

Known Issues:
EnablePaypalExpressCheckoutAndSubmitAnOrderUsingPaypalExpressCheckoutTest : [ACQE-8007](https://jira.corp.adobe.com/browse/ACQE-8007)
